### DISCUSSION TITLE
Cortex.cpp: Updater

### DISCUSSION THREAD

**Discussion by namchuai** [janhq/nitro#1116] [Architecture Specs]
**Motivation**
How should we update cortex binary?
1. Command to update cortex: `cortex update`?
2. Do we use package managers (brew, etc.)
3. How do we do the update? (via script?)
4. Some caveats:
- Cortex can be running as daemon process. So replace it when it's running is not possible (windows for example)
5. Do we support downgrade or running at a specific version?

**Discussion**

**Related**
- https://github.com/janhq/cortex.cpp/issues/1087

**Comment by hiento09** [comment]
Whenever the Cortex CLI is called, a background task checks the latest version from a specific channel (latest release for `stable`, pre-release for `beta`, S3 for `nightly`). If an update is available, it echoes a message to the user's terminal session, such as: 'A new update (version x.y.z) is available. Please run cortex update to update to the latest version.

**Comment by dan-menlo** [comment]
@hiento09 @vansangpfiev Btw, for the Updater UX, I like how Github CLI does it: 

![Image](https://github.com/user-attachments/assets/63c7d933-8a4a-41bb-a8a6-0bbfc5fdaa0f)

**Comment by vansangpfiev** [comment]
@dan-homebrew Thanks for your suggestion. I think we can use that pattern
```
.\cortex.exe -v
v0.4.9

A new release of cortex is available: v0.4.9 -> v0.5.0
To upgrade, run: cortex update
https://github.com/janhq/cortex.cpp/releases/tag/v0.5.0
```

**Comment by hiento09** [comment]
Yeah, this is a nice approach

**Comment by hiento09** [comment]
For nightly channel, binary file will be push to aws s3, the folder structure on s3 will be
```
/cortex
    /latest
        /windows-amd64
            /cortex-nightly.tar.gz
        /linux-amd64
            /cortex-nightly.tar.gz
        /mac-amd64
            /cortex-nightly.tar.gz
        /mac-arm64
            /cortex-nightly.tar.gz
        /version.json
    /vx.y.z
        /windows-amd64
            /cortex-nightly.tar.gz
        /linux-amd64
            /cortex-nightly.tar.gz
        /mac-amd64
            /cortex-nightly.tar.gz
        /mac-arm64
            /cortex-nightly.tar.gz
        /version.json
    ```

**Reply by hiento09** [reply]
cc @vansangpfiev

**Comment by namchuai** [comment]
@vansangpfiev 
nit: `To upgrade, run: cortex update`
Maybe: `To upgrade, run: cortex upgrade` is more verbally correct?
